evm/v0.5/contracts/dev/PrepaidAggregator.sol

```diff
@@ -0,0 +1,435 @@
+    uint128 available = availableFunds.sub(payment);
+
+    availableFunds = available;
+    allocatedFunds = allocatedFunds.add(payment);
+    oracles[msg.sender].withdrawable = oracles[msg.sender].withdrawable.add(payment);
+
+    emit AvailableFundsUpdated(available);
+  }
+
+  function recordSubmission(int256 _answer, uint32 _id)
+    private
+    onlyIfAcceptingAnswers(_id)
+  {
+    rounds[_id].details.answers.push(_answer);
+    oracles[msg.sender].lastReportedRound = _id;
+    oracles[msg.sender].latestAnswer = _answer;
+  }
+
+  function deleteRound(uint32 _id)
+    private
+    onlyIfMaxAnswersReceived(_id)
+  {
+    delete rounds[_id].details;
+  }
+
+  /**
+   * Modifiers
+   */
+
+  modifier onlyValidOracleRound(uint32 _id) {
+    uint32 startingRound = oracles[msg.sender].startingRound;
+    require(startingRound != 0, "Only updatable by whitelisted oracles");
+    require(startingRound <= _id, "New oracles cannot participate in in-progress rounds");
+    require(oracles[msg.sender].endingRound >= _id, "Oracle has been removed from whitelist");
+    require(oracles[msg.sender].lastReportedRound < _id, "Cannot update round reports");
+    _;
+  }
+
+  modifier onlyIfMinAnswersReceived(uint32 _id) {
+    if (rounds[_id].details.answers.length >= rounds[_id].details.minAnswers) {
+      _;
+    }
+  }
+
+  modifier onlyIfMaxAnswersReceived(uint32 _id) {
+    if (rounds[_id].details.answers.length == rounds[_id].details.maxAnswers) {
+      _;
+    }
+  }
+
+  modifier onlyIfAcceptingAnswers(uint32 _id) {
+    require(rounds[_id].details.maxAnswers != 0, "Max responses reached for round");
+    _;
+  }
+
+  modifier onlyOnNewRound(uint32 _id) {
+    if (_id == currentRound.add(1)) {
+      _;
+    }
+  }
+
+  modifier onlyIfDelayed(uint32 _id) {
+    uint256 lastStarted = oracles[msg.sender].lastStartedRound;
+    if (_id > lastStarted + restartDelay) {
+      _;
+    }
+  }
+
+  modifier onlyValidRoundId(uint32 _id) {
+    require(_id == currentRound || _id == currentRound.add(1), "Must report on current round");
+    if (_id > 1) {
+      require(rounds[_id.sub(1)].updatedTimestamp > 0, "Cannot bump round until previous round has an answer");
+    }
+    _;
+  }
+
+  modifier onlyValidRange(uint32 _min, uint32 _max, uint32 _restartDelay) {
+    uint32 oracleNum = oracleCount; // Save on storage reads
+    require(oracleNum >= _max, "Cannot have the answer max higher oracle count");
+    require(_max >= _min, "Cannot have the answer minimum higher the max");
+    if (oracleNum > 0) {
+      require(oracleNum > _restartDelay, "Restart delay must be less than oracle count");
+    }
+    _;
+  }
+
+  modifier onlyUnenabledAddress(address _oracle) {
+    require(oracles[_oracle].endingRound != ROUND_MAX, "Address is already recorded as an oracle");
+    _;
+  }
+
+  modifier onlyEnabledAddress(address _oracle) {
+    require(oracles[_oracle].endingRound == ROUND_MAX, "Address is not a whitelisted oracle");
```

Comment:
Abusing `endingRound` like this is a bit ugly. What if you wanted to enable admin privileges on an oracle for a finite period? Also, the revert message is unclear. `need admin privs to add oracles` might be clearer.

Reply:
There is currently no way to do it for a finite period other than to call `removeOracle` when you want to remove them. Happy to refactor it if we expand the scope to include that functionality, but adding another `enabled` bit raises gas costs and I think the context of this makes it pretty clear how it's being used. I'd prefer to keep it.

I will update the error message to make it more verbose.

Reply:
Upon revisiting this, the modifier is used to check if the sender has permission to update the answer, not to add oracles. Do you still think the error message is unclear?

Reply:
I only see it [on removeOracle](https://github.com/smartcontractkit/chainlink/pull/1807/files/ec3ed098ac271b999189665bb649b429a6386b48#diff-dc7fb26df1fbd9b3bbffc47b6a98b046R150).

Reply:
I see, it is used on remove oracle, but the validation check doesn't have to do with admin privileges. It is checking whether the oracle is currently whitelisted or not. It makes sure you don't remove an address that wasn't currently added. Does the error message make more sense with that context?

Reply:
Oh, I misunderstood the purpose of `endingRound` I thought it prospective, not retrospective. Perhaps the field ought to have a comment about that in the struct.